storage/migrations/0009_oasis_3_emerald_init.up.sql

```diff
@@ -16,10 +19,56 @@ CREATE TABLE IF NOT EXISTS oasis_3.emerald_rounds
   messages_hash    TEXT NOT NULL,
   in_messages_hash TEXT NOT NULL,
 
+  -- TODO: These should also be NOT NULL, but they're populated separately.
+  num_transactions INTEGER,
+  gas_used         BIGINT,
+  size             INTEGER,
+
   -- Arbitrary additional data.
   extra_data JSON
 );
 
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_transactions
+(
+  round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL,
+  tx_hash TEXT NOT NULL,
+  tx_eth_hash TEXT,
+  PRIMARY KEY (round, tx_index)
+);
+
+CREATE INDEX ix_emerald_transactions_tx_hash ON oasis_3.emerald_transactions(tx_hash);
+CREATE INDEX ix_emerald_transactions_tx_eth_hash ON oasis_3.emerald_transactions(tx_eth_hash);
+
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_transaction_signers
+(
+  round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL,
+  signer_index INTEGER NOT NULL,
+  signer_address TEXT NOT NULL,
+  nonce BIGINT NOT NULL,
+  PRIMARY KEY (round, tx_index, signer_index)
+);
+
+CREATE INDEX ix_emerald_transaction_signers_signer_address_signer_nonce ON oasis_3.emerald_transaction_signers(signer_address, nonce);
+
+CREATE TABLE IF NOT EXISTS oasis_3.emerald_related_transactions
+(
+  account_address TEXT NOT NULL,
+  tx_round BIGINT NOT NULL,
+  tx_index INTEGER NOT NULL
+);
+
+CREATE INDEX ix_emerald_related_transactions_address_height_index ON oasis_3.emerald_related_transactions(account_address, tx_round, tx_index);
+
+CREATE TABLE IF NOT EXISTS oasis_1.address_preimages
```

Comment:
nit: maybe place this `CREATE` right under the `oasis_1` schema creation?

Reply:
later changed to be in oasis_3 instead